MeasurementService should provide more guidance about moving version number to the .serviceconfig file

# Bug Report

If the service specifies the version number in code but not in the serviceconfig file, the discovery service assumes that the service is unversioned and assigns it a default version number. If this doesn't match, calling the measurement service from TestStand will fail.

From @nick-beer 

> Yeah - in fact, passing the a version to the constructor is problematic _if_ it is not also found in the .serviceconfig file.  We should probably update the service.py code to enforce this.  The problem is as follows:
>
>- You specify a version in your constructor, but _not_ in your .serviceconfig file.
>- When the service is published to plugin library or statically installed, the DiscoveryService thinks the service is _unversioned_ - because there's no version in the .serviceconfig file.
>- As a result, it registers it internally with a default version (used to be `1.0.0`, but as of last night, it's now `0.0.1`)
>- However, when the service starts itself and registers itself with discovery service, it uses this version that's passed in the constructor, and _those versions don't match_.  This means DiscoveryService fails to see this as a valid registration/resolution, and the client is left hanging, waiting for that valid registration.
>
> ...
> I think an appropriate fix is to allow the following and raise an error for anything else:
>- No version specified in either the constructor or serviceconfig.  This will just get the default version
>- No version specified in the constructor, but a version specified in the serviceconfig.
>- A version specified in the constructor, and that same version is specified in the serviceconfig
>
>If someone has an existing service with a version in the constructor (effectively all services), they'll be required to make some kind of change when updating SDK versions.  This is unfortunate, but because DiscoveryService reads the .serviceconfig file to get version information, I'm not sure how else to get around it.

See this PR discussion for more details:
https://github.com/ni/measurement-plugin-python/pull/907#discussion_r1769231913

[AB#2863530](https://dev.azure.com/ni/94b22d7b-ad7b-4f5e-88f0-867910f91c94/_workitems/edit/2863530)

The test failure looks like flakiness in one of the new client cancellation tests:
```
================================== FAILURES ===================================
____ test___streaming_measurement_execution___cancel___cancels_measurement ____

measurement_plugin_client_module = <module 'test_measurement_client.py' from 'C:\\Users\\ec2-user\\AppData\\Local\\Temp\\pytest-of-unknown\\pytest-5\\measurement_plugin_client_files7\\test_measurement_client.py'>

    def test___streaming_measurement_execution___cancel___cancels_measurement(
        measurement_plugin_client_module: ModuleType,
    ) -> None:
        test_measurement_client_type = getattr(measurement_plugin_client_module, "TestMeasurement")
        measurement_plugin_client = test_measurement_client_type()
    
        with pytest.raises(grpc.RpcError) as exc_info:
            with concurrent.futures.ThreadPoolExecutor() as executor:
                measure = executor.submit(lambda: list(measurement_plugin_client.stream_measure()))
                measurement_plugin_client.cancel()
>               measure.result()
E               Failed: DID NOT RAISE <class 'grpc.RpcError'>

C:\actions-runner\1\_work\measurement-plugin-python\measurement-plugin-python\packages\generator\tests\acceptance\test_streaming_measurement_client.py:96: Failed
```
I filed #910

--- a/packages/service/ni_measurement_plugin_sdk_service/measurement/service.py
+++ b/packages/service/ni_measurement_plugin_sdk_service/measurement/service.py
@@ -219,7 +219,10 @@ class MeasurementService:
             )
 
         if version:
-            warnings.warn("The 'version' parameter is deprecated. Specify the version in the .serviceconfig file instead.", DeprecationWarning)
+            warnings.warn(
+                "The 'version' parameter is deprecated. Specify the version in the .serviceconfig file instead.",
+                DeprecationWarning,
+            )
         config_version = service.get("version")
         if version and config_version is None:
             raise RuntimeError(

--- a/packages/service/ni_measurement_plugin_sdk_service/measurement/service.py
+++ b/packages/service/ni_measurement_plugin_sdk_service/measurement/service.py
@@ -225,9 +225,7 @@ class MeasurementService:
             )
         config_version = service.get("version")
         if version and config_version is None:
-            raise RuntimeError(
-                f"Version error: version parameter is '{version}'. Specify the version using the 'version' field in the .serviceconfig file instead.",
-            )
+            version = None
         if config_version is not None:
             if version and version != config_version:
                 raise RuntimeError(

--- a/packages/service/ni_measurement_plugin_sdk_service/measurement/service.py
+++ b/packages/service/ni_measurement_plugin_sdk_service/measurement/service.py
@@ -218,7 +218,13 @@ class MeasurementService:
                 f"Service class '{service_class}' not found in '{service_config_file}'"
             )
 
+        if version:
+            warnings.warn("The 'version' parameter is deprecated. Specify the version in the .serviceconfig file instead.", DeprecationWarning)
         config_version = service.get("version")
+        if version and config_version is None:
+            raise RuntimeError(
+                f"Version error: version parameter is '{version}'. Specify the version in the .serviceconfig file instead.",
+            )
         if config_version is not None:
             if version and version != config_version:
                 raise RuntimeError(

--- a/packages/service/ni_measurement_plugin_sdk_service/measurement/service.py
+++ b/packages/service/ni_measurement_plugin_sdk_service/measurement/service.py
@@ -220,13 +220,13 @@ class MeasurementService:
 
         if version:
             warnings.warn(
-                "The 'version' parameter is deprecated. Specify the version in the .serviceconfig file instead.",
+                "The 'version' constructor parameter is deprecated. Specify the version using the 'version' field in the .serviceconfig file instead.",
                 DeprecationWarning,
             )
         config_version = service.get("version")
         if version and config_version is None:
             raise RuntimeError(
-                f"Version error: version parameter is '{version}'. Specify the version in the .serviceconfig file instead.",
+                f"Version error: version parameter is '{version}'. Specify the version using the 'version' field in the .serviceconfig file instead.",
             )
         if config_version is not None:
             if version and version != config_version:

--- a/packages/service/ni_measurement_plugin_sdk_service/measurement/service.py
+++ b/packages/service/ni_measurement_plugin_sdk_service/measurement/service.py
@@ -225,7 +225,7 @@ class MeasurementService:
             )
         config_version = service.get("version")
         if version and config_version is None:
-            version = None
+            version = ""
         if config_version is not None:
             if version and version != config_version:
                 raise RuntimeError(

--- a/packages/service/tests/acceptance/test_measurement_service.py
+++ b/packages/service/tests/acceptance/test_measurement_service.py
@@ -253,7 +253,7 @@ def _validate_get_metadata_response(
     ]
 ):
     assert get_metadata_response.measurement_details.display_name == "Loopback Measurement (Py)"
-    assert get_metadata_response.measurement_details.version == "0.1.0.0"
+    assert get_metadata_response.measurement_details.version == "1.2.3.4"
 
     expected_version = "v1"
     if isinstance(get_metadata_response, v2_measurement_service_pb2.GetMetadataResponse):

--- a/packages/service/tests/assets/example.NoVersion.serviceconfig
+++ b/packages/service/tests/assets/example.NoVersion.serviceconfig
@@ -1,0 +1,19 @@
+{
+  "services": [
+    {
+      "displayName": "SampleMeasurement",
+      "serviceClass": "SampleMeasurement_Python",
+      "descriptionUrl": "https://www.example.com/SampleMeasurement.html",
+      "providedInterfaces": [
+        "ni.measurementlink.measurement.v1.MeasurementService",
+        "ni.measurementlink.measurement.v2.MeasurementService"
+      ],
+      "path": "start.bat",
+      "annotations": {
+        "ni/service.description": "Measure inrush current with a shorted load and validate results against configured limits.",
+        "ni/service.collection": "CurrentTests.Inrush",
+        "ni/service.tags": [ "powerup", "current" ]
+      }
+    }
+  ]
+}

--- a/packages/service/tests/assets/example.v1.serviceconfig
+++ b/packages/service/tests/assets/example.v1.serviceconfig
@@ -7,6 +7,7 @@
       "providedInterfaces": [
         "ni.measurementlink.measurement.v1.MeasurementService"
       ],
+      "version": "1.0.1",
       "path": "start.bat"
     }
   ]

--- a/packages/service/tests/unit/test_service.py
+++ b/packages/service/tests/unit/test_service.py
@@ -450,9 +450,9 @@ def test___service_config___create_measurement_service_with_version___service_co
     test_assets_directory: pathlib.Path,
 ):
     measurement_service = MeasurementService(
-            service_config_path=test_assets_directory / "example.NoVersion.serviceconfig",
-            version="2.0.1",
-        )
+        service_config_path=test_assets_directory / "example.NoVersion.serviceconfig",
+        version="2.0.1",
+    )
 
     assert not measurement_service.service_info.versions[0]
 

--- a/packages/service/tests/unit/test_service.py
+++ b/packages/service/tests/unit/test_service.py
@@ -365,7 +365,7 @@ no_annotations: typing.Dict[str, str] = {}
         ),
         (
             "example.v1.serviceconfig",
-            "",
+            "1.0.1",
             ["ni.measurementlink.measurement.v1.MeasurementService"],
             no_annotations,
         ),
@@ -446,11 +446,23 @@ def test___service_config___create_measurement_service_with_version___version_di
     assert "Version mismatch" in str(version_error.value)
 
 
+def test___service_config___create_measurement_service_with_version___service_config_has_no_version(
+    test_assets_directory: pathlib.Path,
+):
+    with pytest.raises(RuntimeError) as version_error:
+        MeasurementService(
+            service_config_path=test_assets_directory / "example.NoVersion.serviceconfig",
+            version="2.0.1",
+        )
+
+    assert "Specify the version in the .serviceconfig file" in str(version_error.value)
+
+
 @pytest.mark.parametrize(
     "service_config,version",
     [
         ("example.serviceconfig", "1.0.1"),
-        ("example.v1.serviceconfig", "2.0.1"),
+        ("example.v1.serviceconfig", "1.0.1"),
     ],
 )
 def test___service_config___create_measurement_service_with_version___version_is_used(

--- a/packages/service/tests/unit/test_service.py
+++ b/packages/service/tests/unit/test_service.py
@@ -455,7 +455,10 @@ def test___service_config___create_measurement_service_with_version___service_co
             version="2.0.1",
         )
 
-    assert "Specify the version using the 'version' field in the .serviceconfig file instead." in str(version_error.value)
+    assert (
+        "Specify the version using the 'version' field in the .serviceconfig file instead."
+        in str(version_error.value)
+    )
 
 
 @pytest.mark.parametrize(

--- a/packages/service/tests/unit/test_service.py
+++ b/packages/service/tests/unit/test_service.py
@@ -455,7 +455,7 @@ def test___service_config___create_measurement_service_with_version___service_co
             version="2.0.1",
         )
 
-    assert "Specify the version in the .serviceconfig file" in str(version_error.value)
+    assert "Specify the version using the 'version' field in the .serviceconfig file instead." in str(version_error.value)
 
 
 @pytest.mark.parametrize(

--- a/packages/service/tests/unit/test_service.py
+++ b/packages/service/tests/unit/test_service.py
@@ -449,16 +449,12 @@ def test___service_config___create_measurement_service_with_version___version_di
 def test___service_config___create_measurement_service_with_version___service_config_has_no_version(
     test_assets_directory: pathlib.Path,
 ):
-    with pytest.raises(RuntimeError) as version_error:
-        MeasurementService(
+    measurement_service = MeasurementService(
             service_config_path=test_assets_directory / "example.NoVersion.serviceconfig",
             version="2.0.1",
         )
 
-    assert (
-        "Specify the version using the 'version' field in the .serviceconfig file instead."
-        in str(version_error.value)
-    )
+    assert not measurement_service.service_info.versions[0]
 
 
 @pytest.mark.parametrize(

--- a/packages/service/tests/utilities/measurements/loopback_measurement/LoopbackMeasurement.serviceconfig
+++ b/packages/service/tests/utilities/measurements/loopback_measurement/LoopbackMeasurement.serviceconfig
@@ -8,6 +8,7 @@
         "ni.measurementlink.measurement.v1.MeasurementService",
         "ni.measurementlink.measurement.v2.MeasurementService"
       ],
+      "version": "1.2.3.4",
       "path": "start.bat",
       "annotations": {
         "ni/service.description": "Measurement plug-in test service that performs a loopback measurement with various data types.",

--- a/packages/service/tests/utilities/measurements/loopback_measurement/__init__.py
+++ b/packages/service/tests/utilities/measurements/loopback_measurement/__init__.py
@@ -20,7 +20,6 @@ class Color(Enum):
 service_directory = pathlib.Path(__file__).resolve().parent
 measurement_service = nims.MeasurementService(
     service_config_path=service_directory / "LoopbackMeasurement.serviceconfig",
-    version="0.1.0.0",
     ui_file_paths=[
         service_directory,
     ],

--- a/packages/service/tests/utilities/measurements/nidaqmx_measurement/__init__.py
+++ b/packages/service/tests/utilities/measurements/nidaqmx_measurement/__init__.py
@@ -11,7 +11,6 @@ from ni_measurement_plugin_sdk_service.session_management import TypedSessionInf
 service_directory = pathlib.Path(__file__).resolve().parent
 measurement_service = nims.MeasurementService(
     service_config_path=service_directory / "NIDAQmxMeasurement.serviceconfig",
-    version="0.1.0.0",
     ui_file_paths=[
         service_directory,
     ],

--- a/packages/service/tests/utilities/measurements/nidcpower_measurement/__init__.py
+++ b/packages/service/tests/utilities/measurements/nidcpower_measurement/__init__.py
@@ -12,7 +12,6 @@ import ni_measurement_plugin_sdk_service as nims
 service_directory = pathlib.Path(__file__).resolve().parent
 measurement_service = nims.MeasurementService(
     service_config_path=service_directory / "NIDCPowerMeasurement.serviceconfig",
-    version="0.1.0.0",
     ui_file_paths=[
         service_directory,
     ],

--- a/packages/service/tests/utilities/measurements/nidigital_measurement/__init__.py
+++ b/packages/service/tests/utilities/measurements/nidigital_measurement/__init__.py
@@ -15,7 +15,6 @@ from ni_measurement_plugin_sdk_service.session_management import (
 service_directory = pathlib.Path(__file__).resolve().parent
 measurement_service = nims.MeasurementService(
     service_config_path=service_directory / "NIDigitalMeasurement.serviceconfig",
-    version="0.1.0.0",
     ui_file_paths=[
         service_directory,
     ],

--- a/packages/service/tests/utilities/measurements/nidmm_measurement/__init__.py
+++ b/packages/service/tests/utilities/measurements/nidmm_measurement/__init__.py
@@ -12,7 +12,6 @@ from ni_measurement_plugin_sdk_service.session_management import TypedSessionInf
 service_directory = pathlib.Path(__file__).resolve().parent
 measurement_service = nims.MeasurementService(
     service_config_path=service_directory / "NIDmmMeasurement.serviceconfig",
-    version="0.1.0.0",
     ui_file_paths=[
         service_directory,
     ],

--- a/packages/service/tests/utilities/measurements/nifgen_measurement/__init__.py
+++ b/packages/service/tests/utilities/measurements/nifgen_measurement/__init__.py
@@ -12,7 +12,6 @@ from ni_measurement_plugin_sdk_service.session_management import TypedSessionInf
 service_directory = pathlib.Path(__file__).resolve().parent
 measurement_service = nims.MeasurementService(
     service_config_path=service_directory / "NIFgenMeasurement.serviceconfig",
-    version="0.1.0.0",
     ui_file_paths=[
         service_directory,
     ],

--- a/packages/service/tests/utilities/measurements/niscope_measurement/__init__.py
+++ b/packages/service/tests/utilities/measurements/niscope_measurement/__init__.py
@@ -12,7 +12,6 @@ from ni_measurement_plugin_sdk_service.session_management import TypedSessionInf
 service_directory = pathlib.Path(__file__).resolve().parent
 measurement_service = nims.MeasurementService(
     service_config_path=service_directory / "NIScopeMeasurement.serviceconfig",
-    version="0.1.0.0",
     ui_file_paths=[
         service_directory,
     ],

--- a/packages/service/tests/utilities/measurements/niswitch_measurement/__init__.py
+++ b/packages/service/tests/utilities/measurements/niswitch_measurement/__init__.py
@@ -11,7 +11,6 @@ from ni_measurement_plugin_sdk_service.session_management import TypedSessionInf
 service_directory = pathlib.Path(__file__).resolve().parent
 measurement_service = nims.MeasurementService(
     service_config_path=service_directory / "NISwitchMeasurement.serviceconfig",
-    version="0.1.0.0",
     ui_file_paths=[
         service_directory,
     ],

--- a/packages/service/tests/utilities/measurements/niswitch_multiplexer_measurement/__init__.py
+++ b/packages/service/tests/utilities/measurements/niswitch_multiplexer_measurement/__init__.py
@@ -11,7 +11,6 @@ from ni_measurement_plugin_sdk_service.session_management import TypedConnection
 service_directory = pathlib.Path(__file__).resolve().parent
 measurement_service = nims.MeasurementService(
     service_config_path=service_directory / "NISwitchMultiplexerMeasurement.serviceconfig",
-    version="0.1.0.0",
     ui_file_paths=[
         service_directory,
     ],

--- a/packages/service/tests/utilities/measurements/pin_aware_measurement/__init__.py
+++ b/packages/service/tests/utilities/measurements/pin_aware_measurement/__init__.py
@@ -8,7 +8,6 @@ import ni_measurement_plugin_sdk_service as nims
 service_directory = pathlib.Path(__file__).resolve().parent
 measurement_service = nims.MeasurementService(
     service_config_path=service_directory / "PinAwareMeasurement.serviceconfig",
-    version="0.1.0.0",
     ui_file_paths=[
         service_directory,
     ],

--- a/packages/service/tests/utilities/measurements/streaming_data_measurement/__init__.py
+++ b/packages/service/tests/utilities/measurements/streaming_data_measurement/__init__.py
@@ -12,7 +12,6 @@ import ni_measurement_plugin_sdk_service as nims
 service_directory = pathlib.Path(__file__).resolve().parent
 measurement_service = nims.MeasurementService(
     service_config_path=service_directory / "StreamingDataMeasurement.serviceconfig",
-    version="0.1.0.0",
     ui_file_paths=[
         service_directory,
     ],

--- a/packages/service/tests/utilities/measurements/unknown_interface_measurement/__init__.py
+++ b/packages/service/tests/utilities/measurements/unknown_interface_measurement/__init__.py
@@ -7,7 +7,6 @@ import ni_measurement_plugin_sdk_service as nims
 service_directory = pathlib.Path(__file__).resolve().parent
 measurement_service = nims.MeasurementService(
     service_config_path=service_directory / "unknown_interface.serviceconfig",
-    version="0.1.0.0",
     ui_file_paths=[
         service_directory,
     ],

--- a/packages/service/tests/utilities/measurements/v1_only_measurement/__init__.py
+++ b/packages/service/tests/utilities/measurements/v1_only_measurement/__init__.py
@@ -7,7 +7,6 @@ import ni_measurement_plugin_sdk_service as nims
 service_directory = pathlib.Path(__file__).resolve().parent
 measurement_service = nims.MeasurementService(
     service_config_path=service_directory / "v1_only.serviceconfig",
-    version="0.1.0.0",
     ui_file_paths=[
         service_directory,
     ],

--- a/packages/service/tests/utilities/measurements/v2_only_measurement/__init__.py
+++ b/packages/service/tests/utilities/measurements/v2_only_measurement/__init__.py
@@ -7,7 +7,6 @@ import ni_measurement_plugin_sdk_service as nims
 service_directory = pathlib.Path(__file__).resolve().parent
 measurement_service = nims.MeasurementService(
     service_config_path=service_directory / "v2_only.serviceconfig",
-    version="0.1.0.0",
     ui_file_paths=[
         service_directory,
     ],

--- a/packages/service/tests/utilities/measurements/yield_vs_return_measurement/__init__.py
+++ b/packages/service/tests/utilities/measurements/yield_vs_return_measurement/__init__.py
@@ -17,7 +17,6 @@ UI_UPDATE_INTERVAL_IN_SECONDS = 100e-3
 service_directory = pathlib.Path(__file__).resolve().parent
 measurement_service = nims.MeasurementService(
     service_config_path=service_directory / "UIProgressUpdates.serviceconfig",
-    version="0.5.0.0",
     ui_file_paths=[
         service_directory,
     ],

